library/init/meta/uneval.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+prelude
+import init.meta.tactic
+
+open tactic
+
+meta class has_uneval (α : Type) :=
+(uneval : α → expr)
+
+@[inline] meta def uneval {α : Type} [has_uneval α] : α → expr :=
+has_uneval.uneval
+
+private meta def c (n : name) := expr.const n []
+
+-- TODO: use macro
+private meta def nat.uneval : nat → expr
+| n :=
+  if n = 0 then c ``nat.zero
+  else
+    let f := if n % 2 = 0 then ``bit0 else ``bit1 in
+    c f (nat.uneval (n / 2))
+
+meta instance : has_uneval nat := ⟨nat.uneval⟩
+
+meta instance : has_uneval char :=
+⟨λ ⟨n, pr⟩, expr.app (expr.const ``char.of_nat []) $ uneval n⟩
```

Comment:
It took me some time to realize that I actually need to convert `vm_obj`s into _unelaborated_ expressions in order to insert them into the tactic expression, so these are different from the `.to_expr` functions (and don't even need `tactic`).